pkg/upgrade/upgrades/alter_sql_instances_binary_version.go

```diff
@@ -0,0 +1,46 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package upgrades
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/catid"
+	"github.com/cockroachdb/cockroach/pkg/upgrade"
+)
+
+const addBinaryVersionColumnToSQLInstancesTableStmt = `
+	ALTER TABLE system.sql_instances
+	ADD COLUMN IF NOT EXISTS binary_version STRING
+	FAMILY "primary"
+`
+
+func alterSystemSQLInstancesTableAddBinaryVersion(
+	ctx context.Context, cs clusterversion.ClusterVersion, d upgrade.TenantDeps,
+) error {
+	tableID := catid.DescID(keys.SQLInstancesTableID)
+	for _, op := range []operation{
+		{
+			name:           "add-sql-instances-binary-version-column",
+			schemaList:     []string{"binary_version"},
+			query:          addBinaryVersionColumnToSQLInstancesTableStmt,
+			schemaExistsFn: columnExists,
+		},
+	} {
+		if err := migrateTable(ctx, cs, d, op, tableID, systemschema.SQLInstancesTable()); err != nil {
```

Comment:
This upgrade isn't actually doing anything in production. The RBR index migration replaces the descriptor with the statically defined descriptor, which would include the column you are adding.

https://github.com/cockroachdb/cockroach/blob/85c6e3874ed6f5e5e47da278b06e26e09d78bfb6/pkg/upgrade/upgrades/system_rbr_indexes.go#L104

Reply:
Should I then remove the migration and keep the test case (run it after RBR index migration)? Anything else I should do? Actually I can't find the version associated with the RBR index migration. Can you please give more details?

Reply:
This is the upgrade and version gate that will replace the descriptor with the new version. https://github.com/cockroachdb/cockroach/blob/ea389ffee029b69ed9f6c5957fd0997cf3f0145a/pkg/upgrade/upgrades/upgrades.go#L250

You should be able to delete your version key and upgrade logic. I think it is a good idea to keep the test. It documents the fact it is an intended side effect of the RBR migration.